src/smtp_observer.py

```diff
@@ -0,0 +1,162 @@
+
+        user = None
+        password = None
+        if relation_data.auth_type == AuthType.PLAIN:
+            user = relation_data.user
+            password = self._get_password_from_relation_data(relation_data)
+
+        # Not all combinations for the next variables are correct. See:
+        # https://github.com/matrix-org/synapse/blob/develop/synapse/config/emailconfig.py
+        force_tls = False
+        enable_tls = False
+        require_transport_security = False
+        if relation_data.transport_security == TransportSecurity.STARTTLS:
+            enable_tls = True
+            require_transport_security = True
+        elif relation_data.transport_security == TransportSecurity.TLS:
+            force_tls = True
+            enable_tls = True
+            require_transport_security = True
+
+        return SMTPConfiguration(
+            enable_tls=enable_tls,
+            force_tls=force_tls,
+            require_transport_security=require_transport_security,
+            host=relation_data.host,
+            port=relation_data.port,
+            user=user,
+            password=password,
+        )
+
+    def _get_password_from_relation_data(self, relation_data: SmtpRelationData) -> Optional[str]:
+        """Get smtp password from relation data.
+
+        Arguments:
+            relation_data: The relation data from where to extract the password
+
+        Returns:
+            the password or None if no password found
+        """
+        # If the relation data password_id exists, that means that
+        # Juju version is >= 3.0 and secrets are used for the password.
+        # Otherwise, use the field password as a fallback
+        if relation_data.password_id:
+            secret = self.model.get_secret(id=relation_data.password_id)
+            content = secret.get_content()
+            return content["password"]
+        return relation_data.password
+
+    def _on_smtp_relation_data_available(self, event: SmtpDataAvailableEvent) -> None:
+        """Handle SMTP data available.
+
+        Arguments:
+            event: The smtp data available event.
+
+        """
+        self.model.unit.status = ops.MaintenanceStatus("Preparing the SMTP integration")
+        logger.debug("_on_smtp_data_available: Enabling SMTP")
+        self._enable_smtp(event)
+
+    def _enable_smtp(self, event: SmtpDataAvailableEvent) -> None:
+        """Enable  SMTP.
+
+        Arguments:
+            event: The event that triggered enabling smtp.
+
+        """
+        container = self._charm.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect() or self._pebble_service is None:
+            self._charm.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
+            # If this event is not deferred, enable_smtp will not be called and
+            # the configuration will not be changed until another event that changes
+            # the configuration is called.
+            event.defer()
```

Comment:
Usually is not advisable to use defer. See [here](https://github.com/canonical/operator/issues/966) and [here](https://juju.is/docs/sdk/deferring-events-details-and-dilemmas).
In this case, since for every configuration change the Synapse is restarted, this might cause unexpected restarts during the charm lifecycle.

Reply:
From what I can see, the `enable_smtp` method is called as part of `src/pebble.py` in `change_config`. `_on_synapse_pebble_ready` calls `src/charm.py`'s `change_config`, and this calls `self.pebble_service.change_config(container)`, so it looks to me like this would happen whenever pebble is ready. Have we observed behaviour that suggests this isn't the case (i.e. if pebble isn't ready when we run `src/smtp_observer.py`'s `_enable_smtp` then `enable_smtp` isn't run as part of the pebble ready hook that fires later)?

Reply:
You are right, it looks like it will work on the next pebble ready. Removed defer call.